File: R/R/path_utils.R
#' Cross-Platform Path Utilities for R MDITRE
#'
#' Provides platform-independent path handling that works seamlessly
#' across Windows, macOS, and Linux/Ubuntu. Works both in development
#' mode and when installed via install.packages().
#'
#' @name path_utils
#' @keywords internal
NULL

#' Get MDITRE Package Installation Directory
#'
#' Returns the directory where the MDITRE R package is installed.
#' This could be in a user library (install.packages()) or a 
#' development directory (devtools::load_all()).
#'
#' @return Character string with absolute path to package directory,
#'   or NULL if package is not installed
#'
#' @examples
#' \dontrun{
#' pkg_dir <- get_package_root()
#' # Installed: "/usr/local/lib/R/site-library/rmditre"
#' # Dev mode: "/home/username/mditre/R"
#' }
#'
#' @export
get_package_root <- function() {
  tryCatch({
    # Try to find package installation directory
    pkg_path <- system.file(package = "rmditre")
    
    if (pkg_path != "") {
      return(normalizePath(pkg_path, winslash = "/", mustWork = FALSE))
    } else {
      # Not installed as package - assume development mode in R/ directory
      current_dir <- getwd()
      if (basename(current_dir) == "R" && dir.exists("R") && file.exists("DESCRIPTION")) {
        # We're in the R package root directory
        return(normalizePath(current_dir, winslash = "/", mustWork = FALSE))
      } else if (file.exists("DESCRIPTION") && file.exists("NAMESPACE")) {
        # Current directory is the package root
        return(normalizePath(current_dir, winslash = "/", mustWork = FALSE))
      } else {
        # Try sourced file location
        return(normalizePath(getwd(), winslash = "/", mustWork = FALSE))
      }
    }
  }, error = function(e) {
    stop(
      "Cannot determine MDITRE package location. ",
      "MDITRE may not be properly installed. ",
      "Please reinstall using: install.packages('mditre')\n",
      "Error details: ", e$message,
      call. = FALSE
    )
  })
}

#' Get MDITRE Project Root Directory (Development Mode Only)
#'
#' Returns the project root directory if in development mode (full
#' repository structure with Python/ and R/ directories). Returns NULL
#' when installed via install.packages().
#'
#' @return Character string with absolute path to project root,
#'   or NULL if in installed mode
#'
#' @examples
#' \dontrun{
#' root <- get_project_root()
#' if (!is.null(root)) {
#'   message("Development mode: ", root)
#' } else {
#'   message("Installed via install.packages()")
#' }
#' }
#'
#' @export
get_project_root <- function() {
  pkg_root <- get_package_root()
  
  if (is.null(pkg_root)) {
    return(NULL)
  }
  
  # Check if we're in development mode
  # Development: /path/to/mditre/R
  # Installed: /path/to/R/library/rmditre
  
  parent_dir <- dirname(pkg_root)
  
  # If parent directory contains both "Python" and "R", we're in dev mode
  if (dir.exists(file.path(parent_dir, "Python")) && 
      dir.exists(file.path(parent_dir, "R"))) {
    return(normalizePath(parent_dir, winslash = "/", mustWork = FALSE))
  }
  
  # Not in development mode
  return(NULL)
}

#' Find Project Root by Looking for Marker Files
#'
#' @param start_dir Starting directory for search
#' @return Path to project root or NULL if not found
#' @keywords internal
find_project_root_by_markers <- function(start_dir) {
  # Marker files that indicate project root
  markers <- c("README.md", "CHANGELOG.md", "Python", "R")
  
  current_dir <- normalizePath(start_dir, winslash = "/", mustWork = FALSE)
  max_levels <- 10  # Prevent infinite loop
  
  for (i in 1:max_levels) {
    # Check if current directory contains markers
    has_markers <- sapply(markers, function(m) {
      file.exists(file.path(current_dir, m)) || 
        dir.exists(file.path(current_dir, m))
    })
    
    # If we find multiple markers, likely the root
    if (sum(has_markers) >= 2) {
      return(current_dir)
    }
    
    # Move up one directory
    parent_dir <- dirname(current_dir)
    if (parent_dir == current_dir) {
      break  # Reached filesystem root
    }
    current_dir <- parent_dir
  }
  
  # Fallback to current working directory
  return(normalizePath(getwd(), winslash = "/", mustWork = FALSE))
}

#' Get Python Implementation Directory (Development Mode Only)
#'
#' Returns the Python/ directory in development mode, or NULL when
#' installed via install.packages().
#'
#' @return Character string with absolute path to Python/ directory,
#'   or NULL if not in development mode
#' @export
get_python_dir <- function() {
  project_root <- get_project_root()
  if (!is.null(project_root)) {
    python_dir <- file.path(project_root, "Python")
    if (dir.exists(python_dir)) {
      return(normalizePath(python_dir, winslash = "/", mustWork = FALSE))
    }
  }
  return(NULL)
}

#' Get R Implementation Directory
#'
#' Returns the R package directory location. In development mode, this is
#' the R/ subdirectory. When installed, this is the package library location.
#'
#' @return Character string with absolute path to R package directory
#' @export
get_r_dir <- function() {
  pkg_root <- get_package_root()
  if (!is.null(pkg_root)) {
    return(pkg_root)
  }
  return(normalizePath(getwd(), winslash = "/", mustWork = FALSE))
}

#' Get Data Directory Path
#'
#' In development mode, returns mditre_root/data/
#' When installed via install.packages(), returns base_path or current directory.
#'
#' @param subdirectory Optional subdirectory within data/ (e.g., 'raw', 'processed')
#' @param base_path Base path for data directory (defaults to project root or cwd)
#' @return Character string with absolute path to data directory
#'
#' @examples
#' \dontrun{
#' # Development mode
#' get_data_dir()  # Returns mditre/data/
#' get_data_dir('raw')  # Returns mditre/data/raw/
#' 
#' # Installed mode - specify your data location
#' get_data_dir(base_path = '/my/data/location')
#' }
#'
#' @export
get_data_dir <- function(subdirectory = NULL, base_path = NULL) {
  if (!is.null(base_path)) {
    data_dir <- base_path
  } else {
    project_root <- get_project_root()
    if (!is.null(project_root)) {
      data_dir <- file.path(project_root, "data")
    } else {
      # Installed via install.packages() - use current directory
      data_dir <- file.path(getwd(), "data")
    }
  }
  
  if (!is.null(subdirectory)) {
    data_dir <- file.path(data_dir, subdirectory)
  }
  return(normalizePath(data_dir, winslash = "/", mustWork = FALSE))
}

#' Get Output Directory for Results
#'
#' In development mode, returns mditre_root/outputs/
#' When installed via install.packages(), returns base_path or current directory.
#'
#' @param create Logical, if TRUE create directory if it doesn't exist
#' @param base_path Base path for output directory (defaults to project root or cwd)
#' @return Character string with absolute path to output directory
#'
#' @examples
#' \dontrun{
#' # Development mode
#' get_output_dir()  # Returns mditre/outputs/
#' 
#' # Installed mode - specify your output location
#' get_output_dir(base_path = '/my/output/location')
#' }
#'
#' @export
get_output_dir <- function(create = TRUE, base_path = NULL) {
  if (!is.null(base_path)) {
    output_dir <- base_path
  } else {
    project_root <- get_project_root()
    if (!is.null(project_root)) {
      output_dir <- file.path(project_root, "outputs")
    } else {
      # Installed via install.packages() - use current directory
      output_dir <- file.path(getwd(), "outputs")
    }
  }
  
  output_dir <- normalizePath(output_dir, winslash = "/", mustWork = FALSE)
  
  if (create && !dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  }
  
  return(output_dir)
}

#' Normalize Path to Platform-Independent Format
#'
#' Converts any path to a platform-independent format using forward slashes.
#'
#' @param path Path to normalize (character string)
#' @param relative_to Optional base path to make relative to
#' @return Normalized path string
#'
#' @examples
#' \dontrun{
#' # Works on all platforms
#' normalize_path("~/data/file.txt")
#' normalize_path("C:\\Users\\data\\file.txt")  # On Windows
#' }
#'
#' @export
normalize_path <- function(path, relative_to = NULL) {
  # Expand ~ and environment variables
  path <- path.expand(path)
  
  # Normalize to platform format, using forward slashes
  path <- normalizePath(path, winslash = "/", mustWork = FALSE)
  
  # Make relative if requested
  if (!is.null(relative_to)) {
    relative_to <- normalizePath(path.expand(relative_to), 
                                 winslash = "/", mustWork = FALSE)
    tryCatch({
      # Try to compute relative path
      path <- gsub(paste0("^", relative_to, "/?"), "", path)
    }, error = function(e) {
      # If can't make relative, return absolute
    })
  }
  
  return(path)
}

#' Ensure Directory Exists
#'
#' Creates a directory and all parent directories if they don't exist.
#'
#' @param directory Directory path to create
#' @return Absolute path to the directory
#'
#' @examples
#' \dontrun{
#' ensure_dir_exists("outputs/models")
#' }
#'
#' @export
ensure_dir_exists <- function(directory) {
  tryCatch({
    directory <- normalizePath(path.expand(directory), 
                              winslash = "/", mustWork = FALSE)
    
    if (!dir.exists(directory)) {
      success <- dir.create(directory, recursive = TRUE, showWarnings = FALSE)
      if (!success) {
        stop("Directory creation failed")
      }
    }
    
    return(directory)
  }, error = function(e) {
    stop(
      "Cannot create directory '", directory, "'. ",
      "Please check directory permissions or choose a different location.\n",
      "Error details: ", e$message,
      call. = FALSE
    )
  })
}

#' Join Path Components
#'
#' Joins multiple path components in a platform-independent way.
#'
#' @param ... Path components to join
#' @return Combined path string
#'
#' @examples
#' \dontrun{
#' join_paths("data", "raw", "file.csv")
#' # Returns "data/raw/file.csv" (platform-independent)
#' }
#'
#' @export
join_paths <- function(...) {
  parts <- list(...)
  if (length(parts) == 0) {
    return("")
  }
  
  path <- file.path(...)
  # Use forward slashes for consistency
  path <- gsub("\\\\", "/", path)
  return(path)
}

#' Get Platform Information
#'
#' Returns information about the current platform.
#'
#' @return Named list with platform details
#'
#' @examples
#' \dontrun{
#' info <- get_platform_info()
#' # Returns list with: os, os_version, r_version, sep, home, cwd
#' }
#'
#' @export
get_platform_info <- function() {
  list(
    os = Sys.info()["sysname"],
    os_version = Sys.info()["release"],
    r_version = paste(R.version$major, R.version$minor, sep = "."),
    sep = .Platform$file.sep,
    pathsep = .Platform$path.sep,
    home = normalizePath(path.expand("~"), winslash = "/", mustWork = FALSE),
    cwd = normalizePath(getwd(), winslash = "/", mustWork = FALSE)
  )
}

#' Convert Path to Unix Style
#'
#' Converts any path to Unix-style with forward slashes.
#' Useful for cross-platform config files.
#'
#' @param path Path to convert
#' @return Path string with forward slashes
#'
#' @examples
#' \dontrun{
#' to_unix_path("C:\\Users\\data\\file.txt")
#' # Returns "C:/Users/data/file.txt"
#' }
#'
#' @export
to_unix_path <- function(path) {
  gsub("\\\\", "/", path)
}

#' Convert Path to Platform-Specific Style
#'
#' Converts path to use platform-appropriate separators.
#'
#' @param path Path to convert
#' @return Path string with platform separators
#'
#' @export
to_platform_path <- function(path) {
  if (.Platform$OS.type == "windows") {
    gsub("/", "\\\\", path)
  } else {
    gsub("\\\\", "/", path)
  }
}

#' Print Platform and Path Information
#'
#' Diagnostic function to display current platform and path configuration.
#'
#' @export
print_path_info <- function() {
  cat("MDITRE Cross-Platform Path Utilities\n")
  cat(strrep("=", 70), "\n\n")
  
  # Platform info
  info <- get_platform_info()
  cat("Platform Information:\n")
  cat(sprintf("  OS: %s %s\n", info$os, info$os_version))
  cat(sprintf("  R Version: %s\n", info$r_version))
  cat(sprintf("  Path Separator: %s\n", info$sep))
  cat(sprintf("  Home Directory: %s\n", info$home))
  cat(sprintf("  Current Directory: %s\n", info$cwd))
  
  # Installation mode
  cat("\n", strrep("=", 70), "\n")
  project_root <- get_project_root()
  if (!is.null(project_root)) {
    cat("Installation Mode: Development (devtools::load_all)\n")
    cat(sprintf("  Project Root: %s\n", project_root))
    python_dir <- get_python_dir()
    if (!is.null(python_dir)) {
      cat(sprintf("  Python Directory: %s\n", python_dir))
    }
    cat(sprintf("  R Directory: %s\n", get_r_dir()))
    cat(sprintf("  Data Directory: %s\n", get_data_dir()))
    cat(sprintf("  Output Directory: %s\n", get_output_dir(create = FALSE)))
  } else {
    cat("Installation Mode: Installed (install.packages)\n")
    cat(sprintf("  Package Location: %s\n", get_package_root()))
    cat(sprintf("  Data Directory: %s (defaults to cwd/data)\n", get_data_dir()))
    cat(sprintf("  Output Directory: %s (defaults to cwd/outputs)\n", get_output_dir(create = FALSE)))
    cat("\nNote: For installed mode, specify data/output paths in your code:\n")
    cat("  get_data_dir(base_path = '/path/to/your/data')\n")
    cat("  get_output_dir(base_path = '/path/to/your/output')\n")
  }
  
  # Path conversion example
  cat("\n", strrep("=", 70), "\n")
  cat("Path Conversion Examples:\n")
  test_path <- "data/raw/file.txt"
  cat(sprintf("  Original: %s\n", test_path))
  cat(sprintf("  Platform: %s\n", to_platform_path(test_path)))
  cat(sprintf("  Unix: %s\n", to_unix_path(test_path)))
  if (!is.null(project_root)) {
    cat(sprintf("  Normalized (abs): %s\n", normalize_path(test_path)))
  }
}

File: Python/mditre/utils/path_utils.py
"""
Cross-platform path utilities for MDITRE.

Provides platform-independent path handling that works seamlessly
across Windows, macOS, and Linux/Ubuntu. Works both in development
mode and when installed via pip.
"""

import os
import sys
from pathlib import Path
from typing import Optional, Union


def get_package_root() -> Path:
    """
    Get the MDITRE package installation directory.

    This returns the location where mditre package is installed,
    which could be in site-packages (pip install) or a development
    directory (pip install -e .).

    Returns:
        Path: Absolute path to the mditre package directory

    Raises:
        RuntimeError: If package location cannot be determined

    Examples:
        >>> root = get_package_root()
        >>> # Pip install: Path('/usr/local/lib/python3.12/site-packages/mditre')
        >>> # Dev install: Path('/home/username/mditre/Python/mditre')
    """
    try:
        # Get the directory containing the mditre package
        return Path(__file__).resolve().parent.parent
    except (NameError, AttributeError) as e:
        raise RuntimeError(
            "Cannot determine MDITRE package location. "
            "MDITRE may not be properly installed. "
            "Please reinstall using: pip install mditre\n"
            f"Error details: {e}"
        ) from e


def get_project_root() -> Optional[Path]:
    """
    Get the MDITRE project root directory (development mode only).

    This only works in development installations (pip install -e .)
    where the full repository structure exists. Returns None when
    installed via pip in site-packages.

    Returns:
        Path or None: Project root if in development mode, None otherwise

    Examples:
        >>> root = get_project_root()
        >>> if root:
        >>>     print(f"Development mode: {root}")
        >>> else:
        >>>     print("Installed via pip")
    """
    # Start from package location
    package_root = get_package_root()

    # Check if we're in development mode (package_root ends with Python/mditre)
    # Development: /path/to/mditre/Python/mditre
    # Pip install: /path/to/site-packages/mditre

    if package_root.parent.name == "Python":
        # We're in development mode
        return package_root.parent.parent
    else:
        # We're in site-packages (pip installed)
        return None


def get_python_dir() -> Path:
    """
    Get the Python implementation directory (development mode only).

    Returns the Python/ directory in development mode, or the package
    directory when installed via pip.

    Returns:
        Path: Python directory path
    """
    project_root = get_project_root()
    if project_root:
        return project_root / "Python"
    else:
        # In pip install mode, return the package root
        return get_package_root().parent


def get_r_dir() -> Optional[Path]:
    """
    Get the R implementation directory (development mode only).

    Only available in development installations. Returns None when
    installed via pip.

    Returns:
        Path or None: R directory if in development mode, None otherwise
    """
    project_root = get_project_root()
    if project_root:
        r_dir = project_root / "R"
        if r_dir.exists():
            return r_dir
    return None
    return get_project_root() / "R"


def get_data_dir(
    subdirectory: Optional[str] = None, base_path: Optional[Union[str, Path]] = None
) -> Path:
    """
    Get the data directory path.

    In development mode, returns mditre_root/data/
    When installed via pip, returns the specified base_path or current directory.

    Args:
        subdirectory: Optional subdirectory within data/ (e.g., 'raw', 'processed')
        base_path: Base path for data directory (defaults to project root or cwd)

    Returns:
        Path: Absolute path to data directory or subdirectory

    Examples:
        >>> # Development mode
        >>> get_data_dir()  # Returns mditre/data/
        >>> get_data_dir('raw')  # Returns mditre/data/raw/
        >>>
        >>> # Pip installed mode - specify your data location
        >>> get_data_dir(base_path='/my/data/location')
    """
    if base_path:
        data_dir = Path(base_path)
    else:
        project_root = get_project_root()
        if project_root:
            data_dir = project_root / "data"
        else:
            # Pip installed - use current working directory
            data_dir = Path.cwd() / "data"

    if subdirectory:
        data_dir = data_dir / subdirectory
    return data_dir


def get_output_dir(create: bool = True, base_path: Optional[Union[str, Path]] = None) -> Path:
    """
    Get the output directory for results.

    In development mode, returns mditre_root/outputs/
    When installed via pip, returns the specified base_path or current directory.

    Args:
        create: If True, create directory if it doesn't exist
        base_path: Base path for output directory (defaults to project root or cwd)

    Returns:
        Path: Absolute path to output directory

    Examples:
        >>> # Development mode
        >>> get_output_dir()  # Returns mditre/outputs/
        >>>
        >>> # Pip installed mode - specify your output location
        >>> get_output_dir(base_path='/my/output/location')
    """
    if base_path:
        output_dir = Path(base_path)
    else:
        project_root = get_project_root()
        if project_root:
            output_dir = project_root / "outputs"
        else:
            # Pip installed - use current working directory
            output_dir = Path.cwd() / "outputs"

    if create and not output_dir.exists():
        output_dir.mkdir(parents=True, exist_ok=True)
    return output_dir


def normalize_path(path: Union[str, Path], relative_to: Optional[Union[str, Path]] = None) -> Path:
    """
    Normalize a path to be platform-independent.

    Args:
        path: Path to normalize (can be string or Path object)
        relative_to: Optional base path to make relative to

    Returns:
        Path: Normalized Path object

    Examples:
        >>> # Convert Windows-style to platform-independent
        >>> normalize_path("C:\\Users\\data\\file.txt")
        Path('C:/Users/data/file.txt')  # On Windows

        >>> # Works on all platforms
        >>> normalize_path("~/data/file.txt")
        Path('/home/username/data/file.txt')  # On Linux
    """
    path = Path(path).expanduser().resolve()

    if relative_to:
        relative_to = Path(relative_to).expanduser().resolve()
        try:
            path = path.relative_to(relative_to)
        except ValueError:
            # Paths are not relative to each other
            pass

    return path


def ensure_dir_exists(directory: Union[str, Path]) -> Path:
    """
    Ensure a directory exists, creating it if necessary.

    Args:
        directory: Directory path to ensure exists

    Returns:
        Path: Absolute path to the directory

    Raises:
        RuntimeError: If directory cannot be created due to permissions or other errors

    Examples:
        >>> ensure_dir_exists("outputs/models")
        Path('/.../mditre/outputs/models')
    """
    try:
        directory = Path(directory).expanduser().resolve()
        directory.mkdir(parents=True, exist_ok=True)
        return directory
    except PermissionError as e:
        raise RuntimeError(
            f"Permission denied: Cannot create directory '{directory}'. "
            "Please check directory permissions or choose a different location.\n"
            f"Error details: {e}"
        ) from e
    except OSError as e:
        raise RuntimeError(
            f"OS error: Cannot create directory '{directory}'. " f"Error details: {e}"
        ) from e


def join_paths(*parts: Union[str, Path]) -> Path:
    """
    Join path parts in a platform-independent way.

    Args:
        *parts: Path components to join

    Returns:
        Path: Combined path

    Examples:
        >>> join_paths("data", "raw", "file.csv")
        Path('data/raw/file.csv')  # Platform-independent
    """
    if not parts:
        return Path()

    base = Path(parts[0])
    for part in parts[1:]:
        base = base / part
    return base


def get_platform_info() -> dict:
    """
    Get current platform information.

    Returns:
        dict: Platform details including OS, separator, home directory

    Examples:
        >>> info = get_platform_info()
        >>> # On Windows: {'os': 'Windows', 'sep': '\\', ...}
        >>> # On macOS:   {'os': 'Darwin', 'sep': '/', ...}
        >>> # On Linux:   {'os': 'Linux', 'sep': '/', ...}
    """
    return {
        "os": sys.platform,
        "os_name": os.name,
        "sep": os.sep,
        "pathsep": os.pathsep,
        "home": str(Path.home()),
        "cwd": str(Path.cwd()),
    }


def to_unix_path(path: Union[str, Path]) -> str:
    """
    Convert path to Unix-style (forward slashes).

    Useful for cross-platform config files or URLs.

    Args:
        path: Path to convert

    Returns:
        str: Path with forward slashes

    Examples:
        >>> to_unix_path("C:\\Users\\data\\file.txt")
        'C:/Users/data/file.txt'
    """
    return str(Path(path)).replace("\\", "/")


def to_platform_path(path: Union[str, Path]) -> str:
    """
    Convert path to platform-specific format.

    Args:
        path: Path to convert

    Returns:
        str: Path with platform-appropriate separators

    Examples:
        >>> # On Windows
        >>> to_platform_path("data/raw/file.txt")
        'data\\raw\\file.txt'

        >>> # On Unix
        >>> to_platform_path("data/raw/file.txt")
        'data/raw/file.txt'
    """
    return str(Path(path))


# Example usage and testing
if __name__ == "__main__":
    print("MDITRE Cross-Platform Path Utilities")
    print("=" * 70)

    # Show platform info
    info = get_platform_info()
    print(f"\nPlatform: {info['os']}")
    print(f"OS Name: {info['os_name']}")
    print(f"Path Separator: {info['sep']}")
    print(f"Home Directory: {info['home']}")
    print(f"Current Directory: {info['cwd']}")

    # Detect installation mode
    print("\n" + "=" * 70)
    project_root = get_project_root()
    if project_root:
        print("Installation Mode: Development (editable install)")
        print(f"Project Root: {project_root}")
        print(f"Python Directory: {get_python_dir()}")
        r_dir = get_r_dir()
        if r_dir:
            print(f"R Directory: {r_dir}")
        print(f"Data Directory: {get_data_dir()}")
        print(f"Output Directory: {get_output_dir(create=False)}")
    else:
        print("Installation Mode: Pip installed (site-packages)")
        print(f"Package Location: {get_package_root()}")
        print(f"Data Directory: {get_data_dir()} (defaults to cwd/data)")
        print(f"Output Directory: {get_output_dir(create=False)} (defaults to cwd/outputs)")
        print("\nNote: For pip-installed mode, specify data/output paths in your code:")
        print("  get_data_dir(base_path='/path/to/your/data')")
        print("  get_output_dir(base_path='/path/to/your/output')")

    # Show path conversions
    print("\n" + "=" * 70)
    print("Path Conversions:")
    test_path = "data/raw/file.txt"
    print(f"  Original: {test_path}")
    print(f"  Platform: {to_platform_path(test_path)}")
    print(f"  Unix: {to_unix_path(test_path)}")
    if project_root:
        print(f"  Normalized (abs): {normalize_path(test_path)}")
